File: sdks/ruby/lib/pushfy/webhooks.rb
# frozen_string_literal: true

require "openssl"

module Pushfy
  # Verifies the authenticity of an incoming Pushfy webhook.
  #
  # Signature header differs by product:
  #   - Messaging status   -> X-Pushfy-Signature: sha256=<hex>   (scheme: :prefixed)
  #   - Push Notifications -> X-Push-Signature:   sha256=<hex>   (scheme: :prefixed)
  #   - Conversational AI  -> X-PA-Signature:      <hex>          (scheme: :raw)
  #
  # Always pass the RAW request body (the exact bytes received), not a
  # re-serialized object — re-serialization changes the signature.
  module Webhooks
    module_function

    # Returns true when the signature is valid. `scheme` is :prefixed (default)
    # or :raw (for X-PA-Signature).
    def verify(payload:, signature:, secret:, scheme: :prefixed)
      return false if signature.nil? || signature.to_s.empty?
      return false if secret.nil? || secret.to_s.empty?

      hex = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha256"), secret, payload.to_s)
      expected = scheme.to_s == "raw" ? hex : "sha256=#{hex}"
      secure_compare(expected, signature.to_s)
    end

    # Convenience wrappers per product.
    def messaging(payload:, signature:, secret:)
      verify(payload: payload, signature: signature, secret: secret, scheme: :prefixed)
    end

    def push(payload:, signature:, secret:)
      verify(payload: payload, signature: signature, secret: secret, scheme: :prefixed)
    end

    def conversations(payload:, signature:, secret:)
      verify(payload: payload, signature: signature, secret: secret, scheme: :raw)
    end

    # Constant-time string comparison (manual, à la Rack::Utils.secure_compare).
    # Returns false on length mismatch without early-outing on content.
    def secure_compare(a, b)
      a = a.to_s.b
      b = b.to_s.b
      return false unless a.bytesize == b.bytesize

      result = 0
      a.bytes.each_with_index { |byte, i| result |= byte ^ b.getbyte(i) }
      result.zero?
    end
  end
end

File: sdks/python/pushfy/webhooks.py
"""Verify the authenticity of incoming Pushfy webhooks.

Signature header differs by product::

    - Messaging status   -> X-Pushfy-Signature: sha256=<hex>   (scheme: 'prefixed')
    - Push Notifications  -> X-Push-Signature:   sha256=<hex>   (scheme: 'prefixed')
    - Conversational AI   -> X-PA-Signature:      <hex>          (scheme: 'raw')

Always pass the RAW request body (the exact bytes received), not a
re-serialized object -- re-serialization changes the signature.
"""

import hashlib
import hmac as _hmac


def verify(payload, signature, secret, scheme="prefixed"):
    """Verify a webhook signature.

    :param payload: Raw request body (``str`` or ``bytes``).
    :param signature: Value of the signature header.
    :param secret: Your webhook secret.
    :param scheme: ``'prefixed'`` (``sha256=<hex>``) or ``'raw'`` (bare hex,
        used by ``X-PA-Signature``).
    :returns: ``True`` when the signature is valid.
    """
    if not signature or not secret:
        return False
    body = payload if isinstance(payload, (bytes, bytearray)) else str(payload).encode("utf-8")
    hex_digest = _hmac.new(
        secret.encode("utf-8"), bytes(body), hashlib.sha256
    ).hexdigest()
    expected = hex_digest if scheme == "raw" else "sha256=" + hex_digest
    return _hmac.compare_digest(expected, str(signature))


def messaging(payload, signature, secret):
    """Verify a messaging status/DLR webhook (``X-Pushfy-Signature``, prefixed)."""
    return verify(payload, signature, secret, scheme="prefixed")


def push(payload, signature, secret):
    """Verify a Push Notifications webhook (``X-Push-Signature``, prefixed)."""
    return verify(payload, signature, secret, scheme="prefixed")


def conversations(payload, signature, secret):
    """Verify a Conversational AI webhook (``X-PA-Signature``, raw hex)."""
    return verify(payload, signature, secret, scheme="raw")
